### DISCUSSION TITLE
PYNQ Python code: How to use AXI-Stream to deliver Vector-Type variable?

### DISCUSSION THREAD

**Discussion by b07901181** [bol-edu/HLS-SOC-Discussions#31] [Q&A]
Hi. I am using AXI Stream interface to transmit a vector type data. As in the following code, I pack a vector in **hls::axis** and then I pack it in **hls::stream**.
**********
typedef signed int int32_t;
typedef hls::vector<int32_t, 8> int32_vt;
typedef hls::axis<int32_vt,1,1,1> value_t;
typedef hls::stream<value_t> stream_t;
**********
I have passed CSim, CSynth, CoSim, and used Vivado to generate the .bit and .hwh files. (I have also checked that the TLAST signal is correct.) But **I don't know how to write the Python code** to run it on FPGA.
I have tried to use **np.int32 Buffer** or the buffer below:
********
dt = np.dtype((np.int32, 8))
Buffer = allocate(shape=(NumSample,), dtype=dt)
********
But both of them have the same problem that they cannot pass **recvchannel.wait()**, as the following error shows:
********
![image](https://user-images.githubusercontent.com/65820999/196042863-42a8410a-359c-400b-ba9d-2dd487a79c00.png)
********
Does anyone know how to solve this?
Thanks.

**Comment by jiinlai** [comment]
Refer to: https://docs.xilinx.com/r/en-US/pg021_axi_dma/MM2S_DMASR-MM2S-DMA-Status-Register-Offset-04h
From the listing of dma.py, there 3 possible errors:
![image](https://user-images.githubusercontent.com/17397765/196590462-c025c4ea-4656-427b-982d-40ae3bcec416.png)

See if it provides any clues.

Jiin

**Comment by jiinlai** [comment]
See if you can print the returned error code, it may give you a better idea of which error is.

Jiin

**Comment by jiinlai** [comment]
I think the hang may be caused by the amount of return data being less than the buffer size. Refer to the ppt below. Please check the loop count for the return data.
![image](https://user-images.githubusercontent.com/17397765/196973474-d1e73b2b-4d22-4275-980d-7594aa0be828.png)
